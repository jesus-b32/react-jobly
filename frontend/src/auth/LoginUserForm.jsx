import React, { useState, useContext } from 'react';
import {
  Form,
  Input,
  Button,
  Label,
  FormGroup,
  Col,
  Container,
  Alert,
} from 'reactstrap';
import { useNavigate, Navigate } from 'react-router-dom';
import UserContext from './UserContext';

function LoginUserForm({ login }) {
  const navigate = useNavigate();
  const { currentUser } = useContext(UserContext);

  /**redirect to homepage if already logged in */
  if (currentUser) return <Navigate to={'/'} />;

  const INITIAL_STATE = {
    username: '',
    password: '',
  };

  const [formData, setFormData] = useState(INITIAL_STATE);
  const [formErrors, setFormErrors] = useState([]);

  function handleChange(event) {
    const { name, value } = event.target;
    setFormData((formData) => ({
      ...formData,
      [name]: value,
    }));
  }

  /** On submit, authenticate user on backend and display any errors.
   *  if successful
   *  - redirect to homepage
   */
  async function handleSubmit(event) {
    event.preventDefault();
    const result = await login(formData);
    if (result.success) {
      navigate('/');
    } else {
      setFormErrors(result.errors);
    }
  }
  return (
    <Container>
      <Form onSubmit={handleSubmit} className='mb-4'>
        <FormGroup row>
          <Label for='username' lg='1'>
            Username:
          </Label>
          <Col lg='10'>
            <Input
              id='username'
              name='username'
              type='text'
              placeholder='username'
              value={formData.username}
              onChange={handleChange}
              className='mb-4'
              required
            />
          </Col>
        </FormGroup>
        <FormGroup row>
          <Label for='password' lg='1'>
            Password:
          </Label>
          <Col lg='10'>
            <Input
              id='password'
              name='password'
              type='password'
              placeholder='password'
              value={formData.password}
              onChange={handleChange}
              className='mb-4'
              required
            />
          </Col>
        </FormGroup>
        {formErrors.length ? <Alert color='danger'>{formErrors}</Alert> : null}
        <Button color='primary'>Submit</Button>
      </Form>
    </Container>
  );
}

export default LoginUserForm;
