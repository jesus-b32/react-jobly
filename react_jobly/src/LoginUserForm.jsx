import React from 'react';

function LoginUserForm() {
  return <h1>LoginUserForm</h1>;
}

export default LoginUserForm;
