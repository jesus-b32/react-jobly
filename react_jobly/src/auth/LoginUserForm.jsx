import React from 'react';
import { useState } from 'react';
import {
  Form,
  Input,
  Button,
  Label,
  FormGroup,
  Col,
  Container,
  FormFeedback,
} from 'reactstrap';
import { useNavigate } from 'react-router-dom';

function LoginUserForm({ login }) {
  const navigate = useNavigate();

  const INITIAL_STATE = {
    username: '',
    password: '',
  };

  const [formData, setFormData] = useState(INITIAL_STATE);

  const [formErrors, setFormErrors] = useState([]);

  function handleChange(event) {
    const { name, value } = event.target;
    setFormData((formData) => ({
      ...formData,
      [name]: value,
    }));
  }

  async function handleSubmit(event) {
    event.preventDefault();
    const result = await login(formData);
    if (result.success) {
      setFormData(INITIAL_STATE);
      navigate('/');
    } else {
      setFormErrors(result.errors);
    }
  }
  return (
    <Container>
      <Form onSubmit={handleSubmit} className='mb-4'>
        <FormGroup row>
          <Label for='username' md='1'>
            Username:
          </Label>
          <Col md='10'>
            <Input
              id='username'
              name='username'
              type='text'
              placeholder='username'
              value={formData.username}
              onChange={handleChange}
              className='mb-4'
              required
            />
          </Col>
        </FormGroup>
        <FormGroup row>
          <Label for='password' md='1'>
            Password:
          </Label>
          <Col md='10'>
            <Input
              id='password'
              name='password'
              type='password'
              placeholder='password'
              value={formData.password}
              onChange={handleChange}
              className='mb-4'
              required
            />
          </Col>
        </FormGroup>
        {formErrors.length ? (
          <FormFeedback tooltip>{formErrors}</FormFeedback>
        ) : null}
        <Button>Submit</Button>
      </Form>
    </Container>
  );
}

export default LoginUserForm;
