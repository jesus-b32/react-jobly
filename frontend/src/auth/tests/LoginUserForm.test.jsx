import LoginUserForm from '../LoginUserForm';
import { MemoryRouter } from 'react-router-dom';
import { it, expect, vi, afterEach } from 'vitest';
import { render, fireEvent, cleanup, screen } from '@testing-library/react';
import UserContext from '../../auth/UserContext';

/**mock a successful login */
const login = vi.fn().mockResolvedValue({ success: true });

const TestLoginUserForm = () => {
  const currentUser = null; //not signed in user
  return (
    <MemoryRouter>
      <UserContext.Provider value={{ currentUser }}>
        <LoginUserForm login={login} />
      </UserContext.Provider>
    </MemoryRouter>
  );
};

it('should render without crashing', () => {
  render(<TestLoginUserForm />);
});

it('should match snapshot', () => {
  const { asFragment } = render(<TestLoginUserForm />);

  expect(asFragment()).toMatchSnapshot();
});

it('should update username field on change', () => {
  render(<TestLoginUserForm />);

  const usernameInput = screen.getByLabelText('Username:');

  // Simulate a user typing into the input field
  fireEvent.change(usernameInput, {
    target: { value: 'johnDoe' },
  });

  // Assert the input value has changed
  expect(usernameInput.value).toBe('johnDoe');
});

it('should login user on successful submit', () => {
  render(<TestLoginUserForm />);

  const usernameInput = screen.getByLabelText('Username:');
  const passwordInput = screen.getByLabelText('Password:');
  const submitBtn = screen.getByText('Submit');

  // Simulate a user typing into the input field
  fireEvent.change(usernameInput, {
    target: { value: 'johnDoe' },
  });
  fireEvent.change(passwordInput, {
    target: { value: 'password' },
  });
  fireEvent.click(submitBtn);

  expect(login).toHaveBeenCalled(1);
  expect(login).toHaveBeenCalledWith({
    username: 'johnDoe',
    password: 'password',
  });
});

afterEach(cleanup);
