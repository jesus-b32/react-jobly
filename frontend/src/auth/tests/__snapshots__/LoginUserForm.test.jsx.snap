// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`should match snapshot 1`] = `
<DocumentFragment>
  <div
    class="container"
  >
    <form
      class="mb-4"
    >
      <div
        class="row mb-3"
      >
        <label
          class="col-lg-1 col-form-label"
          for="username"
        >
          Username:
        </label>
        <div
          class="col-lg-10"
        >
          <input
            class="mb-4 form-control"
            id="username"
            name="username"
            placeholder="username"
            required=""
            type="text"
            value=""
          />
        </div>
      </div>
      <div
        class="row mb-3"
      >
        <label
          class="col-lg-1 col-form-label"
          for="password"
        >
          Password:
        </label>
        <div
          class="col-lg-10"
        >
          <input
            class="mb-4 form-control"
            id="password"
            name="password"
            placeholder="password"
            required=""
            type="password"
            value=""
          />
        </div>
      </div>
      <button
        class="btn btn-primary"
      >
        Submit
      </button>
    </form>
  </div>
</DocumentFragment>
`;
